### PR TITLE
Require the RegisterBlock associated with a serial::Instance to implement RegisterBlockImpl

### PR DESCRIPTION
**Pull request by georgehosono**
The `Serial::new` method cannot be called using a generic `serial::Instance` because the `Instance` trait does not require that the associated `RegisterBlock` implements `RegisterBlockImpl`. See the example below.

```rust
use stm32f4xx_hal::{
    gpio::PushPull,
    prelude::*,
    rcc::Clocks,
    serial::{Instance, Serial},
    uart::config::InvalidConfig,
};

struct SerialConnection<UartX: Instance>(Serial<UartX, u8>);

fn make_serial<UartX: Instance>(
    usart: UartX,
    tx: impl Into<UartX::Tx<PushPull>>,
    rx: impl Into<UartX::Rx<PushPull>>,
    baud_rate: u32,
    clocks: &Clocks,
) -> Result<SerialConnection<UartX>, InvalidConfig> {
    Serial::<_, u8>::new(usart, (tx, rx), baud_rate.bps(), &clocks)
        .map(|conn| SerialConnection(conn))
}
```
This fails to compile with the following error.
```
error[E0277]: the trait bound `<UartX as stm32f4xx_hal::serial::Instance>::RegisterBlock: stm32f4xx_hal::serial::uart_impls::RegisterBlockImpl` is not satisfied
   --> src/main.rs:18:26
    |
18  |     Serial::<_, u8>::new(usart, (tx, rx), baud_rate.bps(), &clo...
    |     -------------------- ^^^^^ the trait `stm32f4xx_hal::serial::uart_impls::RegisterBlockImpl` is not implemented for `<UartX as stm32f4xx_hal::serial::Instance>::RegisterBlock`
    |     |
    |     required by a bound introduced by this call
    |
note: required by a bound in `Serial::<USART, WORD>::new`
   --> /home/george/.cargo/registry/src/index.crates.io-6f17d22bba15001f/stm32f4xx-hal-0.20.0/src/serial.rs:233:45
    |
223 | ...fn new(
    |       --- required by a bound in this associated function
...
233 | ...<USART as Instance>::RegisterBlock: uart_impls::RegisterBlockIm...
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Serial::<USART, WORD>::new`
help: consider further restricting the associated type
    |
17  | ) -> Result<SerialConnection<UartX>, InvalidConfig> where <UartX as stm32f4xx_hal::serial::Instance>::RegisterBlock: stm32f4xx_hal::serial::uart_impls::RegisterBlockImpl {
    |                                                     +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

Every `Instance` has a `RegisterBlock` that implements `RegisterBlockImpl`, so we can simply require it. This also makes many trait bounds unneeded. There are many impl block that require `where <USART as Instance>::RegisterBlock: RegisterBlockImpl,`. But since `RegisterBlock` always implements `RegisterBlockImpl`, those lines are unneeded.

Let me know if I've misunderstood something. I think this is fine, but I'm still figuring out how this crate works.

Thank you!

### CODE REVIEW

**Comment by burrbull** [comment]
cc @qwerty19106 could you also look at this?

**Review by burrbull — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,15 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
...
  - Check features enabled in `build-rs` [#718]
+  - add trait bound `RegisterBlockImpl` to type `RegisterBlock` associated with `serial::Instance` [#732]
+  - removed unneeded trait bound for methods that take in a `serial::Instance` and use the associated `RegisterBlock`
+
+### Changed
+
+ - complete and rework Dma Stream API [#666]
+ - SPI bidi takes 2 pins [#526]
+ - `Fast Read Quad I/O (EBh)` in `qspi-w25q` example now matches W25QXX datasheet. [#682]
+ - `embedded-storage` version bumped to 0.3
```
Comment (burrbull) [inline CHANGELOG.md:29]: move changelog in `Unreleased` section, please. And remove unrelated changes.

Comment (burrbull) [reply CHANGELOG.md:29]: and squash commits

Comment (georgehosono) [reply CHANGELOG.md]: These should both be fixed in the lastest commit. Not sure how I messed up the changelog. Thanks for catching that.

**Comment by qwerty19106** [comment]
This is a huge simplification. I didn't know that this syntax is possible.
Thanks for your work!
